pkg/operation/common/utils.go

```diff
@@ -547,3 +550,107 @@ func CurrentReplicaCount(client client.Client, namespace, deploymentName string)
+// maintenance time window of the Shoot.
+func IsNowInEffectiveShootMaintenanceTimeWindow(shoot *gardenv1beta1.Shoot) bool {
+	return EffectiveShootMaintenanceTimeWindow(shoot).Contains(time.Now())
+}
+
+// IsOperationOngoing checks whether the Shoot's generation has changed or if the LastOperation status
+// is not Succeeded.
+func IsOperationOngoing(shoot *gardenv1beta1.Shoot) bool {
+	lastOperation := shoot.Status.LastOperation
+	return shoot.Generation != shoot.Status.ObservedGeneration ||
+		(lastOperation != nil && lastOperation.State != gardencorev1alpha1.LastOperationStateSucceeded)
+}
+
+// SyncPeriodOfShoot determines the sync period of the given shoot.
+//
+// If no overwrite is possible, the defaultMinSyncPeriod is returned.
+// Otherwise, the overwrite is parsed. If an error occurs or it is smaller than the defaultMinSyncPeriod,
+// the defaultMinSyncPeriod is returned. Otherwise, the overwrite is returned.
+func SyncPeriodOfShoot(respectSyncPeriodOverwrite bool, defaultMinSyncPeriod time.Duration, shoot *gardenv1beta1.Shoot) time.Duration {
+	if !RespectShootSyncPeriodOverwrite(respectSyncPeriodOverwrite, shoot) {
+		return defaultMinSyncPeriod
+	}
+
+	syncPeriodOverwrite, ok := shoot.Annotations[ShootSyncPeriod]
+	if !ok {
+		return defaultMinSyncPeriod
+	}
+
+	syncPeriod, err := time.ParseDuration(syncPeriodOverwrite)
+	if err != nil {
+		return defaultMinSyncPeriod
+	}
+
+	if syncPeriod < defaultMinSyncPeriod {
+		return defaultMinSyncPeriod
+	}
+	return syncPeriod
+}
+
+// EffectiveMaintenanceTimeWindow cuts a maintenance time window at the end with a guess of 15 minutes. It is subtracted from the end
+// of a maintenance time window to use a best-effort kind of finishing the operation before the end.
+// Generally, we can't make sure that the maintenance operation is done by the end of the time window anyway (considering large
+// clusters with hundreds of nodes, a rolling update will take several hours).
+func EffectiveMaintenanceTimeWindow(timeWindow *utils.MaintenanceTimeWindow) *utils.MaintenanceTimeWindow {
+	return timeWindow.WithEnd(timeWindow.End().Add(0, -15, 0))
+}
+
+// EffectiveShootMaintenanceTimeWindow returns the effective MaintenanceTimeWindow of the given Shoot.
+func EffectiveShootMaintenanceTimeWindow(shoot *gardenv1beta1.Shoot) *utils.MaintenanceTimeWindow {
+	maintenance := shoot.Spec.Maintenance
+	if maintenance == nil || maintenance.TimeWindow == nil {
+		return utils.AlwaysTimeWindow
+	}
+
+	timeWindow, err := utils.ParseMaintenanceTimeWindow(maintenance.TimeWindow.Begin, maintenance.TimeWindow.End)
+	if err != nil {
+		return utils.AlwaysTimeWindow
+	}
+
+	return EffectiveMaintenanceTimeWindow(timeWindow)
+}
+
+// DurationUntilNextShootSync computes the duration until the next sync of a Shoot.
+func DurationUntilNextShootSync(respectSyncPeriodOverwrite bool, defaultMinSyncPeriod time.Duration, shoot *gardenv1beta1.Shoot) time.Duration {
```

Comment:
This function doesn't respect the configuration whether maintenance should happen only in reconcile time window or not. Unless `now` is within the shoot's time window it always computes the next sync time to be randomly in the time window.

Please add a check such that if `reconcileInMaintenanceOnly=false` it doesn't care about the time window.